snafu-derive/src/lib.rs

```diff
@@ -203,67 +210,91 @@ impl fmt::Display for DuplicateAttribute {
     /// `location` is a string describing where the attribute can be specified, often referencing a
     /// parent element, for example "on a field".
     fn new(name: &'static str, location: &'static str) -> Self {
         Self {
             name,
             location,
-            value: None,
+            values: Vec::new(),
             errors: SyntaxErrors::default(),
         }
     }
 
-    /// Add an occurence of the attribute found at the given `span`.
-    fn add(&mut self, item: T, span: proc_macro2::Span) {
-        if self.value.is_some() {
+    /// Add an occurence of the attribute found at the given token tree `tts`.
+    fn add(&mut self, item: T, tts: U) {
+        if !self.values.is_empty() {
             self.errors.add(
-                span,
+                tts.clone(),
                 DuplicateAttribute {
                     attribute: self.name,
                     location: self.location,
                 },
             );
-        } else {
-            self.value = Some(item);
         }
+        self.values.push((item, tts));
     }
 
     #[allow(dead_code)]
     /// Returns the number of elements that have been added.
     fn len(&self) -> usize {
-        if self.value.is_some() {
-            1 + self.errors.len()
-        } else {
-            0
-        }
+        self.values.len()
     }
 
     /// Returns true if no elements have been added, otherwise false.
+    #[allow(dead_code)]
     fn is_empty(&self) -> bool {
-        self.value.is_none()
+        self.values.is_empty()
+    }
+
+    /// Returns an iterator over all values that have been added.
+    ///
+    /// This can help with additional manual error checks beyond the duplication checks that
+    /// `AtMostOne` handles for you.
+    fn iter(&self) -> std::slice::Iter<(T, U)> {
+        self.values.iter()
     }
 
     /// Consumes the AtMostOne, returning the first item added, if any, and the list of errors
     /// representing any items added beyond the first.
     fn finish(self) -> (Option<T>, Vec<syn::Error>) {
+        let (value, errors) = self.finish_with_location();
+        (value.map(|(val, _location)| val), errors)
+    }
+
+    /// Like `finish` but also returns the location of the first item added.  Useful when you have
+    /// to do additional, manual error checking on the first item added, and you'd like to report
+    /// an accurate location for it in case of errors.
+    fn finish_with_location(mut self) -> (Option<(T, U)>, Vec<syn::Error>) {
         let errors = match self.errors.finish() {
             Ok(()) => Vec::new(),
             Err(vec) => vec,
         };
-        (self.value, errors)
+        // We're done with the `AtMostOne` and want to return the first element, if any.
+        // Truncate to throw away any extras, and pop to get an Option with (maybe) the first.
+        self.values.truncate(1);
+        (self.values.pop(), errors)
```

Comment:
In the realm of potentially micro-optimizing, you could use a `VecDeque` here, but I wouldn't worry about it ;-)

Reply:
Ah, regardless of optimization, it's clearer for the purpose!  Switching to `VecDeque`.